File: VB/CalculatedPropertiesSolution.Module/BusinessObjects/Product.vb
Imports DevExpress.Data.Filtering
Imports DevExpress.Persistent.Base
Imports DevExpress.Persistent.BaseImpl
Imports DevExpress.Xpo
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks

Namespace CalculatedPropertiesSolution.Module.BusinessObjects
    <DefaultClassOptions> _
    Public Class Product
        Inherits BaseObject

        Public Sub New(ByVal session As Session)
            MyBase.New(session)
        End Sub
        Private fName As String
        Public Property Name() As String
            Get
                Return fName
            End Get
            Set(ByVal value As String)
                SetPropertyValue("Name", fName, value)
            End Set
        End Property
        <Association("Product-Orders"), Aggregated> _
        Public ReadOnly Property Orders() As XPCollection(Of Order)
            Get
                Return GetCollection(Of Order)("Orders")
            End Get
        End Property
        Private fOrdersCount? As Integer = Nothing
        Public ReadOnly Property OrdersCount() As Integer?
            Get
                If Not IsLoading AndAlso Not IsSaving AndAlso fOrdersCount Is Nothing Then
                    UpdateOrdersCount(False)
                End If
                Return fOrdersCount
            End Get
        End Property
        Private fOrdersTotal? As Decimal = Nothing
        Public ReadOnly Property OrdersTotal() As Decimal?
            Get
                If Not IsLoading AndAlso Not IsSaving AndAlso fOrdersTotal Is Nothing Then
                    UpdateOrdersTotal(False)
                End If
                Return fOrdersTotal
            End Get
        End Property
        Private fMaximumOrder? As Decimal = Nothing
        Public ReadOnly Property MaximumOrder() As Decimal?
            Get
                If Not IsLoading AndAlso Not IsSaving AndAlso fMaximumOrder Is Nothing Then
                    UpdateMaximumOrder(False)
                End If
                Return fMaximumOrder
            End Get
        End Property
        Public Sub UpdateOrdersCount(ByVal forceChangeEvents As Boolean)
            Dim oldOrdersCount? As Integer = fOrdersCount
            fOrdersCount = Convert.ToInt32(Evaluate(CriteriaOperator.Parse("Orders.Count")))
            If forceChangeEvents Then
                OnChanged("OrdersCount", oldOrdersCount, fOrdersCount)
            End If
        End Sub
        Public Sub UpdateOrdersTotal(ByVal forceChangeEvents As Boolean)
            Dim oldOrdersTotal? As Decimal = fOrdersTotal
            Dim tempTotal As Decimal = 0D
            For Each detail As Order In Orders
                tempTotal += detail.Total
            Next detail
            fOrdersTotal = tempTotal
            If forceChangeEvents Then
                OnChanged("OrdersTotal", oldOrdersTotal, fOrdersTotal)
            End If
        End Sub
        Public Sub UpdateMaximumOrder(ByVal forceChangeEvents As Boolean)
            Dim oldMaximumOrder? As Decimal = fMaximumOrder
            Dim tempMaximum As Decimal = 0D
            For Each detail As Order In Orders
                If detail.Total > tempMaximum Then
                    tempMaximum = detail.Total
                End If
            Next detail
            fMaximumOrder = tempMaximum
            If forceChangeEvents Then
                OnChanged("MaximumOrder", oldMaximumOrder, fMaximumOrder)
            End If
        End Sub
        Protected Overrides Sub OnLoaded()
            Reset()
            MyBase.OnLoaded()
        End Sub
        Private Sub Reset()
            fOrdersCount = Nothing
            fOrdersTotal = Nothing
            fMaximumOrder = Nothing
        End Sub
    End Class
End Namespace

File: CS/CalculatedPropertiesSolution.Module/BusinessObjects/Product.cs
using DevExpress.Data.Filtering;
using DevExpress.Persistent.Base;
using DevExpress.Persistent.BaseImpl;
using DevExpress.Xpo;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace CalculatedPropertiesSolution.Module.BusinessObjects {
    [DefaultClassOptions]
    public class Product : BaseObject {
        public Product(Session session) : base(session) { }
        private string fName;
        public string Name {
            get { return fName; }
            set { SetPropertyValue("Name", ref fName, value); }
        }
        [Association("Product-Orders"), Aggregated]
        public XPCollection<Order> Orders {
            get { return GetCollection<Order>("Orders"); }
        }
        private int? fOrdersCount = null;
        public int? OrdersCount {
            get {
                if (!IsLoading && !IsSaving && fOrdersCount == null)
                    UpdateOrdersCount(false);
                return fOrdersCount;
            }
        }
        private decimal? fOrdersTotal = null;
        public decimal? OrdersTotal {
            get {
                if (!IsLoading && !IsSaving && fOrdersTotal == null)
                    UpdateOrdersTotal(false);
                return fOrdersTotal;
            }
        }
        private decimal? fMaximumOrder = null;
        public decimal? MaximumOrder {
            get {
                if (!IsLoading && !IsSaving && fMaximumOrder == null)
                    UpdateMaximumOrder(false);
                return fMaximumOrder;
            }
        }
        public void UpdateOrdersCount(bool forceChangeEvents) {
            int? oldOrdersCount = fOrdersCount;
            fOrdersCount = Convert.ToInt32(Evaluate(CriteriaOperator.Parse("Orders.Count")));
            if (forceChangeEvents)
                OnChanged("OrdersCount", oldOrdersCount, fOrdersCount);
        }
        public void UpdateOrdersTotal(bool forceChangeEvents) {
            decimal? oldOrdersTotal = fOrdersTotal;
            decimal tempTotal = 0m;
            foreach (Order detail in Orders)
                tempTotal += detail.Total;
            fOrdersTotal = tempTotal;
            if (forceChangeEvents)
                OnChanged("OrdersTotal", oldOrdersTotal, fOrdersTotal);
        }
        public void UpdateMaximumOrder(bool forceChangeEvents) {
            decimal? oldMaximumOrder = fMaximumOrder;
            decimal tempMaximum = 0m;
            foreach (Order detail in Orders)
                if (detail.Total > tempMaximum)
                    tempMaximum = detail.Total;
            fMaximumOrder = tempMaximum;
            if (forceChangeEvents)
                OnChanged("MaximumOrder", oldMaximumOrder, fMaximumOrder);
        }
        protected override void OnLoaded() {
            Reset();
            base.OnLoaded();
        }
        private void Reset() {
            fOrdersCount = null;
            fOrdersTotal = null;
            fMaximumOrder = null;
        }
    }
}
